lib/plugin/GitBase.js

```diff
@@ -50,11 +50,20 @@ class GitBase extends Plugin {
     return !_.includes(remoteUrlOrName, '/');
   }
 
-  getRemoteUrl() {
-    const remoteNameOrUrl = this.options.pushRepo || 'origin';
+  async getRemoteUrl() {
+    const remoteNameOrUrl = this.options.pushRepo || (await this.getBranchUpstreamRemote()) || 'origin';
     return this.isRemoteName(remoteNameOrUrl)
       ? this.exec(`git config --get remote.${remoteNameOrUrl}.url`, { options }).catch(() => null)
-      : Promise.resolve(remoteNameOrUrl);
+      : remoteNameOrUrl;
+  }
+
+  async getBranchUpstreamRemote() {
+    let remote = null;
+    const branchName = await this.exec('git rev-parse --abbrev-ref HEAD', { options });
```

Comment:
Please reuse `this.getBranchName`

Reply:
I would need to move `getBranchName` to this class, I believe, since GitBase can be extended by other classes than `Git` and then `getBranchName` wouldn't be available.

Reply:
BTW. It's kinda unclear what should be in Git and what in GitBase. I would guess that GitBase should have all pure-git stuff that is not related to release-it functionality itself but then Git has methods like `this.hasUpstreamBranch()` and `this.tagExists()` that, according to that understanding, would fit better in GitBase.

Reply:
Done some changes. Let me know if that's what you had in mind. You can be picky about naming and all as I was contemplating those a bit.

Reply:
Thanks, this is what I had in mind.

Regarding the classes. Everything is now a plugin, so all release-it does is executing methods on those (i.e. there is no "release-it functionality"). GitBase is the base class for the Git plugin, but also the base for the GitRelease class, which in turn is the base class for both the GitLab and Github plugins. I just wanted to put common functionality in a base class for reuse, I guess.

Reply:
What I had in mind by saying "release-it functionality" is for example: `hasUpstreamBranch` or `tagExists` being pure-git functionality, versus `isRequiredBranch` being related to the package's functionality. I would think that former two could be in GitBase, if following that logic.

Reply:
Right, I see. One could argue that e.g. `isRequiredBranch` and `hasUpstreamBranch` is specific Git-plugin functionality, as the other plugins don't need it. E.g. when doing a GitHub release, this is not used/required.

Btw, I now see that `tagExists` is not even used at all, so I can remove it :)